### PR TITLE
Add missing docs for the vm module

### PR DESCRIPTION
**Pull request by qinsoon**
This PR is a step towards https://github.com/mmtk/mmtk-core/issues/309.
* Add some tests and documents to clarify `HeaderMetadata.bit_offset` (related discussion: https://mmtk.zulipchat.com/#narrow/stream/315620-Porting/topic/ScalaNative.2FMMTK/near/398587245).
* Modify the macro `define_vm_metadata_spec!` to allow adding docs for the generated types.
* Add missing docs for public items in the `vm` module.

Please feel free to make edits to the PR if there is any issue.

### CODE REVIEW

File: src/vm/scanning.rs [hunk 1]
```diff
@@ -242,6 +242,14 @@ pub trait Scanning<VM: VMBinding> {
     /// Return whether the VM supports return barriers. This is unused at the moment.
     fn supports_return_barrier() -> bool;
 
+    /// Prepare for another round of root scanning in the same GC. Some GC algorithms
+    /// need multiple transitive closures, and each transitive closure starts from
+    /// root scanning. We expect the binding to provide the same root set for every
+    /// round of root scanning in the same GC. Bindings can use this call to get
+    /// ready for another round of root scanning to make sure that the same root
+    /// set will be returned in the upcoming calls of root scannig methods,
```
Comment (k-sareen) [inline src/vm/scanning.rs:250]:
```suggestion
    /// set will be returned in the upcoming calls of root scanning methods,
```

File: src/vm/object_model.rs [hunk 2]
```diff
@@ -488,17 +488,33 @@ pub mod specs {
...
-        ($spec_name: ident, $is_global: expr, $log_num_bits: expr, $side_min_obj_size: expr) => {
+        ($(#[$outer:meta])*$spec_name: ident, $is_global: expr, $log_num_bits: expr, $side_min_obj_size: expr) => {
+            $(#[$outer])*
             pub struct $spec_name(MetadataSpec);
             impl $spec_name {
+                #[doc="The number of bits (in log2) that are needed for the spec."]
                 pub const LOG_NUM_BITS: usize = $log_num_bits;
+
+                #[doc="Whether this spec is global or local. The binding needs to make sure"]
+                #[doc="global specs are laid out after a global spec, and local specs are laid"]
```
Comment (wks) [inline src/vm/object_model.rs:499]: I think the notion of "laid out after" only applies for side metadata.  In header metadata only takes a offset, therefore the order doesn't matter.  We should mention "side metadata" here.

File: src/vm/mod.rs [hunk 3]
```diff
@@ -19,6 +19,7 @@ use crate::util::constants::*;
 
 mod active_plan;
 mod collection;
+/// Allows MMTk to access edges in a VM defined way.
```
Comment (k-sareen) [inline src/vm/mod.rs:22]:
```suggestion
/// Allows MMTk to access edges in a VM-defined way.
```

**Review by wks — CHANGES_REQUESTED** [review]

File: src/vm/object_model.rs [hunk 4]
```diff
@@ -488,17 +488,33 @@ pub mod specs {
     // This macro is invoked in define_vm_metadata_global_spec or define_vm_metadata_local_spec.
     // Use those two to define a new VM metadata spec.
     macro_rules! define_vm_metadata_spec {
-        ($spec_name: ident, $is_global: expr, $log_num_bits: expr, $side_min_obj_size: expr) => {
+        ($(#[$outer:meta])*$spec_name: ident, $is_global: expr, $log_num_bits: expr, $side_min_obj_size: expr) => {
+            $(#[$outer])*
             pub struct $spec_name(MetadataSpec);
             impl $spec_name {
+                #[doc="The number of bits (in log2) that are needed for the spec."]
```
[the remark this answers is not in the document]
Comment (qinsoon) [reply src/vm/object_model.rs]: Yeah. It works fine. I have changed most `#[doc]` to `///`.

File: src/vm/object_model.rs (hunk 2 above)
Comment (qinsoon) [reply src/vm/object_model.rs]: Done

**Review by wks — APPROVED** [review]
LGTM

**Pull request merged**